Refine non-convex README section (#183)
qpsolvers/qpsolvers: README.md

\end{split}
$$

> Vector inequalities apply coordinate by coordinate. The function returns the solution $x^\*$ found by the solver, or ``None`` in case of failure/unfeasible problem. For all the solvers, the matrix $P$ should be [positive  semi-definite](https://en.wikipedia.org/wiki/Definite_symmetric_matrix) which implies a convex problem. For some solvers the more strict of condition of positive definiteness is required.

Reviewer: I agree it's an improvement to mention both :+1: 

Suggesting one more iteration:
```suggestion
Vector inequalities apply coordinate by coordinate. The function returns the solution $x^\*$ found by the solver, or ``None`` in case of failure/unfeasible problem. All solvers require the problem to be convex, meaning the matrix $P$ should be [positive semi-definite](https://en.wikipedia.org/wiki/Definite_symmetric_matrix). Some solvers further require the problem to be strictly convex, meaning $P$ should be positive definite.
```

- *I have a squared norm in my cost function, how can I apply a QP solver to my problem?*
  - You can [cast squared norms to QP matrices](https://scaron.info/blog/conversion-from-least-squares-to-quadratic-programming.html) and feed the result to [`solve_qp`](https://qpsolvers.github.io/qpsolvers/quadratic-programming.html#qpsolvers.solve_qp).
- *I have a non-convex quadratic program. Is there a solver I can use?*
  - Unfortunately most available QP solvers are designed for convex problems.

>   - If your cost matrix *P* is semi-definite rather than definite, try OSQP.

Author: This seems out of place to me. Most of "serious" QP solvers support both strictly and non-strictly convex problems. Why explicitly mention OSQP? Even more so, why mention about strict convexity in the non-convex section? In terms of convergence rate of the methods, there is very little lost between strictly-convex <-> non-strictly-convex problems as compared to convex <-> non-convex.

>   - If your cost matrix *P* is semi-definite rather than definite, try OSQP.

Reviewer: Yeah, this point is definitely not still valid :see_no_evil: I checked again in for instance the script below, and the only solver that in itself requires strictly convex is ``quadprog``.

Thanks for pushing back on this, :dart: again.

```python
import numpy as np

from qpsolvers import available_solvers, solve_qp

n = 500
P = np.zeros((n, n))
q = np.ones(n)
lb = -np.ones(n)
ub = +np.ones(n)

for solver in available_solvers:
    try:
        x = solve_qp(P, q, lb=lb, ub=ub, solver=solver)
    except Exception:
        print(f"{solver=}: failure")
```

>   - Unfortunately most available QP solvers are designed for convex problems (i.e. problems for which `P` is positive semidefinite). That's in a way expected, as solving non-convex QP problems is NP hard.

Reviewer: ```suggestion
  - Unfortunately most available QP solvers are designed for convex problems (i.e. problems for which $P$ is positive semi-definite).
```

Solving convex QP problems is [also NP-hard](https://cs.stackexchange.com/a/115549).

Author: It's not - [algorithms with polynomial complexity exist since the 80s that solve convex convex QPs](https://link.springer.com/article/10.1007/BF01587086). It's unclear to me what the link you shared says about (continuous as opposed to mixed-integer) convex QPs.

Reviewer: Sorry for my cursory feedback on this. You're right, interior point methods exist in polynomial time.

We could even add "... is NP hard, whereas solving convex QPs can be done in polynomial time." here.

>   - CPLEX has methods for solving non-convex quadratic problems to either local or global optimality. Notice that [finding global solutions](https://www.gurobi.com/wp-content/uploads/2020-01-14_Non-Convex-Quadratic-Optimization-in-Gurobi-9.0-Webinar.pdf?x58432) can be significantly slower than [finding local solutions](https://link.springer.com/chapter/10.1007/978-1-4613-0263-6_8).

Reviewer: Can we link to CPLEX's functionality directly, e.g. in the doc, or slides as with Gurobi?

The second sentence is a pertinent point to make. Can we link it to a general discussion of global vs local optimality in non-convex QPs? E.g. in terms of complexity rather than individual methods. (Also, a link to non-paywalled material would be more accessible.) Maybe there is *e.g.* a section in Boyd & Vandenberghe we could link to?

Alternatively, I think no link would be OK too. It should be already intuitive that global optimization is harder than local, perhaps we don't want to go too deep in the FAQ.

Author: > Maybe there is e.g. a section in Boyd & Vandenberghe we could link to?

As far as I recall, not, they only deal with convex problems.

> It should be already intuitive that global optimization is harder than local, perhaps we don't want to go too deep in the FAQ.

In convex optimisation it's not: every local optimiser is also a global optimiser. This repo only supports convex optimisation, so when diverting from the "convex world" I think it's nice to remind some things that are lost.

Reviewer: I found the following page for CPLEX: [Distinguishing between convex and nonconvex QPs](https://www.ibm.com/docs/en/icos/12.9.0?topic=qp-distinguishing-between-convex-nonconvex-qps).

Here is the corresponding suggestion:

```suggestion
  - CPLEX has methods for solving non-convex quadratic problems to [either local or global optimality](https://www.ibm.com/docs/fr/icos/22.1.0?topic=qp-distinguishing-between-convex-nonconvex-qps). Notice that finding global solutions can be significantly slower than [finding local solutions](https://link.springer.com/chapter/10.1007/978-1-4613-0263-6_8).
```

It also removes the duplicate link to Gurobi, which we have just below. That page from the CPLEX docs also explains the global/local distinction nicely.
